subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -1119,7 +1025,43 @@ public ServiceProvider getService(LookupContext context, TypeSpec serviceType) {
             return null;
         }
 
-        private ServiceProvider wrap(final Object instance) {
+        @Override
+        public void getAll(LookupContext context, Class<?> serviceType, List<ServiceProvider> result) {
+            if (parent instanceof DefaultServiceRegistry) {
+                ((DefaultServiceRegistry) parent).collectInto(serviceType, new IsolatedServiceList(result));
+                return;
+            }
+            List<?> services = parent.getAll(serviceType);
+            assert services != null : String.format("parent returned null for services of type %s", format(serviceType));
+            for (Object service : services) {
+                result.add(instanceToServiceProvider(service));
+            }
+        }
+
+        private class IsolatedServiceList extends AbstractList<ServiceProvider> {
```

Comment:
:o: What is the meaning of "isolated" here? It's not a static inner class, so it keeps a reference to the service registry, just to get the parent.

Reply:
It's isolating services coming from a parent registry so we don't leak dependency references back into the parent. I'll add some documentation around here.